sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLogSuite.scala

```diff
@@ -186,4 +190,57 @@ class HDFSMetadataLogSuite extends SparkFunSuite with SharedSQLContext {
     intercept[AssertionError](verifyBatchIds(Seq(1), Some(2L), Some(1L)))
     intercept[AssertionError](verifyBatchIds(Seq(0), Some(2L), Some(1L)))
   }
+
+  test("HDFSMetadataLog: retry") {
+    withTempDir { temp =>
+      spark.conf.set(
+        SQLConf.STREAMING_CHECKPOINT_FILE_MANAGER_CLASS.parent.key,
+        classOf[FakeFileSystemBasedCheckpointFileManager].getName)
+      spark.sessionState.conf.setConfString(SQLConf.META_DATA_NUM_RETRIES.key, 1.toString)
+      val metadataLog1 = new HDFSMetadataLog[String](spark, temp.getAbsolutePath)
+      intercept[IOException] {
+        assert(metadataLog1.add(0, "batch"))
+      }
+
+      spark.sessionState.conf.setConfString(
+        SQLConf.META_DATA_NUM_RETRIES.key, SQLConf.META_DATA_NUM_RETRIES.defaultValue.get.toString)
+      val metadataLog2 = new HDFSMetadataLog[String](spark, temp.getAbsolutePath)
+      assert(metadataLog2.add(1, "batch1"))
+      assert(metadataLog2.get(1) === Some("batch1"))
+    }
+  }
+}
+
+class FakeFileSystemBasedCheckpointFileManager(path: Path, conf: Configuration)
```

Comment:
Didn't check all the details but is it not possible to use `CreateAtomicTestManager`?

Reply:
Yeah, it can be but also need create a sub class.